app/connect/sep/page.tsx

```diff
@@ -3,19 +3,39 @@ import { authRequestLink } from "@/app/_lib/radar/rest-source/service";
 import { withBasePath } from "@/app/_lib/util/links";
 import crypto from "crypto";
 import { useSearchParams } from "next/navigation";
+import { useEffect } from "react";
+
+const RETURN_TO_STORAGE_PATH = 'connect_sep_return_to'
+const STATE_STORAGE_PATH = 'connect_sep_state'
 
 export default function Page() {
-  const projectId = localStorage.getItem('studyId')
-  const code = useSearchParams().get('code')
-  const state = crypto.randomBytes(20).toString('hex')
-  
-  if (code == null) {
-    window.location.replace(authRequestLink(state))
-  } else {
-    let searchParams = new URLSearchParams([['code', code]])
-    fetch(withBasePath('/api/connect/sep?' + searchParams.toString())).then(
-      (r) => window.location.replace(withBasePath(`/${projectId}/portal`))
-    )
-  }
+  const searchParams = useSearchParams()
+  const code = searchParams.get('code')
+  const returnedState = searchParams.get('state')
+
+  useEffect(() => {
+    if (code == null) {
+      const state = crypto.randomBytes(20).toString('hex')
+      const return_to = searchParams.get('return_to')
+      localStorage.setItem(RETURN_TO_STORAGE_PATH, return_to ?? '/')
+      localStorage.setItem(STATE_STORAGE_PATH, state)
```

Comment:
Do we need to store state even if connect/sep and connect/armt are separate?

Reply:
The state is stored so that we can compare the state sent in the request with the returned value, to ensure no CSRF (https://stackoverflow.com/questions/26132066/what-is-the-purpose-of-the-state-parameter-in-oauth-authorization-request)

Reply:
Oh okay I see, thanks for clarifying. 👍